File: Guias_praticos/Guia02/Java/MergeSort.java
package mergesort;
import java.io.*;
import java.util.Random;
import java.util.concurrent.TimeUnit;

class MergeSort
{
    public void Merge(int[] A, int p, int q, int r){
        int n1 = q - p + 1;
        int n2 = r - q;
        int[] L = new int[n1+1];
        int[] R = new int[n2+1];
        for(int i = 0; i < n1; i++) L[i] = A[p + i];
        for(int j = 0; j < n2; j++) R[j] = A[q + j + 1];
        L[n1] = Integer.MAX_VALUE; R[n2] = Integer.MAX_VALUE;
        int i = 0, j = 0;
        for(int k = p; k <= r; k++)
            if(L[i] <= R[j])
                A[k] = L[i++];
            else
                A[k] = R[j++];
     }
    
    public void run(int[] A, int p, int r){
        if(p < r){
            int q = (int) Math.floor((p+r) / 2.0);
            run(A, p, q); run(A, q+1, r); Merge(A, p, q, r);
        }
    }
 
    static void printArray(int arr[])
    {
        int n = arr.length;
        for (int i=0; i<n; ++i)
            System.out.print(arr[i] + " ");
        System.out.println();
    }
    
    public static double calculoDesvio(double tempos[], int tamanho)
    {
        double soma = 0.0, desvioPadrao = 0.0;

        for(double num : tempos) {
            soma += num;
        }

        double media = soma / tamanho;

        for(double num: tempos) {
            desvioPadrao += Math.pow(num - media, 2);
        }

        return Math.sqrt(desvioPadrao / tamanho);
    }
    
    public static void main(String args[]) throws IOException
    {
        try (Writer writer = new BufferedWriter(new OutputStreamWriter(
              new FileOutputStream(
                      "C:\\Algoritmos\\Java\\MergeSort\\MergeSort_Java.txt"),
                "utf-8"))) {
            String newline = System.getProperty("line.separator");
            
            int numListas = 50;
            int tamanho = 0;
            int iteracoes = 31;
			int incremento = 25000;

            for (int z = 1; z <= iteracoes; z++){
                double[] tempos;
                tempos = new double[numListas];
                float somaTempos = 0;
                int contador = 1;            

                for (int k = 0; k < numListas; k++){
                    int[] lista;
                    lista = new int[tamanho];

                    Random rand = new Random(); 

                    for (int i = 0; i < tamanho; ++i) {      
                        lista[i] = rand.nextInt(1000);
                    }

                    System.out.println("Lista "+ contador +" de tamanho "
                            + tamanho);
                    //printArray(lista);            

                    MergeSort ob = new MergeSort();
                    long tempoInicial = System.nanoTime(); 
                    ob.run(lista, 0, lista.length-1);
                    long tempoFinal = System.nanoTime() - tempoInicial;

                    /*
                    System.out.println("Lista "+ k +" de tamanho "+ tamanho +
                                        " organizada");
                    printArray(lista);
                    */

                    // Tempo em MiliSegundos
                    float tempoMS = TimeUnit.NANOSECONDS.toMillis(tempoFinal); 
                    // Tempo em Segundos
                    float tempoS = (float)tempoMS / 1000;  

                    System.out.println("Tempo de execucao :: "+ tempoMS +"ms");
                    System.out.println("Tempo de execucao :: "+ tempoS +"s");
                    System.out.println();
                    somaTempos += tempoS;
                    tempos[k] = tempoS;
                    contador++;
                }

                float media = somaTempos / numListas;
                double desvio = calculoDesvio(tempos, tamanho);
                System.out.println("Media para "+ numListas +" listas de "+ 
                        tamanho + " elementos :: " + media + "ms");
                System.out.println("Desvio para "+ numListas +" listas de "+
                        tamanho + " elementos :: " + 
                        desvio + "ms");
                System.out.println();
                String s = (tamanho + " " + media + " " + desvio + newline);
                writer.write(s);
                tamanho += incremento;
            }
            
            writer.close();
        }
    }
}

File: Guias_praticos/Guia02/C/MergeSort.c
// MergeSort.cpp : Defines the entry point for the console application.
//
#include "stdafx.h"
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>
#define numListas 2

void merge(int lista[], int esq, int m, int drt)
{
	int i, j, k;
	int n1 = m - esq + 1;
	int n2 = drt - m;

	/* Arrays temporárias */	
	int *L = n1;
	int *R = n2;
	L = (int*) malloc(sizeof(int) * n1);
	R = (int*) malloc(sizeof(int) * n2);
	
	for (i = 0; i < n1; i++)
		L[i] = lista[esq + i];
	for (j = 0; j < n2; j++)
		R[j] = lista[m + 1 + j];

	/* Merge das arrays temporárias com a lista */
	i = 0;
	j = 0;
	k = esq;
	while (i < n1 && j < n2)
	{
		if (L[i] <= R[j])
		{
			lista[k] = L[i];
			i++;
		}
		else
		{
			lista[k] = R[j];
			j++;
		}
		k++;
	}

	while (i < n1)	{
		lista[k] = L[i];
		i++;
		k++;
	}
	while (j < n2)	{
		lista[k] = R[j];
		j++;
		k++;
	}
	free(L);
	free(R);
}

void mergeSort(int lista[], int esq, int drt)
{
	if (esq < drt)
	{
		// O mesmo que (esq+drt)/2, mas evita overflow para números muito
		// grandes em "esq" e "drt"
		int m = esq + (drt - esq) / 2;

		// Sort da primeira e segunda metades
		mergeSort(lista, esq, m);
		mergeSort(lista, m + 1, drt);

		merge(lista, esq, m, drt);
	}
}

void printArray(int lista[], int size)
{
	for (int i = 0; i < size; i++)
		printf("%d ", lista[i]);
}

int main()
{
	int tamanho = 0;
	int iteracoes = 31;
	int incremento = 5000;
	FILE *f;
	f = fopen("C:\\Algoritmos\\C\\MergeSort.txt", "w+");
	srand(time(NULL));

	for (int j = 0; j < iteracoes; j++) {
		int ciclo = 1;
		int num;
		/* Variáveis usadadas no cálculo da média e desvio padrão */
		float soma = 0;
		float variancia;
		float somaVar = 0;
		float somaTempos = 0;
		float tempos[numListas];

		printf("\nITERACAO n.%d", j + 1);

		for (int k = 0; k < numListas; k++) {
			int *lista;
			lista = malloc(sizeof(int) * tamanho);
			for (int i = 0; i < tamanho; i++) {
				num = rand() % 1000;
				lista[i] = num;
			}
			
			printf("\n\nLista nao-organizada de tamanho %d ::\n", tamanho);
			printArray(lista, tamanho);
			
			printf("\n");
			clock_t inicio = clock();
			mergeSort(lista, 0, tamanho - 1);
			clock_t fim = clock();
			printf("Lista n.%d de tamanho %d ::", ciclo, tamanho);
			
			printf("\n");
			printArray(lista, tamanho);
			printf("\n");
			
			float tempo = ((float)(fim - inicio) / CLOCKS_PER_SEC);
			somaTempos += tempo;
			tempos[k] = tempo;
			printf(" %f s", tempo);
			ciclo++;
			free(lista);
		}
		float media = somaTempos / (float)numListas;
		for (int i = 0; i < numListas; i++)		{
			somaVar = somaVar + pow((tempos[i] - media), 2);
		}
		variancia = somaVar / (float)numListas;
		float desvioP = sqrt(variancia);

		printf("\n\nMedia para %d listas de %d elementos :: %f s\n", numListas, tamanho, media);
		printf("Desvio para %d listas de %d elementos :: %f s\n", numListas, tamanho, desvioP);
		fprintf(f, "%d %f %f\n", tamanho, media, desvioP);
		tamanho += incremento;
	}
	fclose(f);
	printf("\nPrograma terminado. Pressione qualquer tecla para fechar a janela.\n");
	getchar();
	return 0;
}
